### PR TITLE
Fix SwathDefinition get_bbox_lonlats returning counter-clockwise coordinates

### PR DESCRIPTION
**Pull request by djhoese**
This is a continuation of #385. That PR originally started as a "my data isn't clockwise so things are incorrect". I then found out that the fake data I was using was actually incorrect and didn't match the real world data I was trying to work with. So switching to 64-bit floats fixed that issue.

Now I'm working with other real world data (MiRS algorithm output) where the lon/lat arrays have their first element (row 0, column 0) in the south-west corner which results in the `get_bbox_lonlats` producing coordinates in a counter-clockwise path. This PR fixes this issue in a very similar way as my original fix #385 except for a few important points:

1. I now only use the first two pixels to determine if the first and second side are in increasing in value. This should get around some of the concerns @mraspaud had from #385.
2. I discovered that the way I was using the utility `create_test_longitude` was actually not allowed. That is, requesting longitudes in decreasing order resulted in the function thinking we were crossing the anti-meridian. This is now fixed in this PR.

I also discovered that `get_edge_lonlats` was depending on the non-clockwise path of the data or at least a test it had was checking for that. So I added the `force_clockwise` keyword argument and set it to False for `get_edge_lonlats` as this makes sure the lonlats follow the data rather than provide a clockwise bbox.

 - [x] Tests added 
 - [x] Tests passed 
 - [x] Passes ``git diff origin/main **/*py | flake8 --diff`` 
 - [x] Fully documented

### CODE REVIEW

**Comment by djhoese** [comment]
I've also added a small patch to the DynamicAreaDefinition. I discovered that dask `nanmin/nanmax` fails if one of the chunks provided to it is empty. This can happen with the way we were using boolean arrays: `np.nanmin(xarr[xarr >= 0])`

File: pyresample/geometry.py [hunk 1]
```diff
@@ -957,6 +978,10 @@ def compute_domain(self, corners, resolution=None, shape=None):
                        corners[1] - y_resolution / 2,
                        corners[2] + x_resolution / 2,
                        corners[3] + y_resolution / 2)
+        # area_extent = (corners[0] - x_resolution,
+        #                corners[1] - y_resolution,
+        #                corners[2] + x_resolution,
+        #                corners[3] + y_resolution)
```
Comment (mraspaud) [inline pyresample/geometry.py:984]: Why the comment here?

File: pyresample/geometry.py [hunk 2]
```diff
@@ -298,6 +308,17 @@ def get_bbox_lonlats(self) -> tuple:
                            (s4_lon.squeeze(), s4_lat.squeeze())])
         if hasattr(lons[0], 'compute') and da is not None:
             lons, lats = da.compute(lons, lats)
+        # compare the first two pixels in the right column
+        lat_is_increasing = lats[1][0] < lats[1][1]
+        # compare the first two pixels in the "top" column
+        lon_is_increasing = lons[0][0] < lons[0][1]
+        is_ccw = (lon_is_increasing and lat_is_increasing) or (not lon_is_increasing and not lat_is_increasing)
```
Comment (mraspaud) [inline pyresample/geometry.py:315]: I can see two issues with this approach:

1. This assumes two things: a. the top and right edges of the swath are perpendicular, and b. the swath is never aligned with the equator. For b., it would indeed be a very strange orbit, so I think we can live with this assumption (although it should be mentioned at least in the code), while a. could be a bit problematic with sensors that are not our usual imagers. An alternative test could be to check that the oriented angle around the top-right corner of the swath is not reflex.
2. pole or antimeridian situations. A solution could be to convert the three points to cartesian and use the angle check described above.

File: pyresample/geometry.py [hunk 3]
```diff
@@ -269,9 +269,19 @@ def get_boundary_lonlats(self):
...
         """Return the bounding box lons and lats.
 
+        Args:
+            force_clockwise:
+                Perform minimal checks and reordering of coordinates to ensure
+                that the returned coordinates follow a clockwise direction.
+                This is important for compatibility with
+                :class:`pyresample.spherical.AreaBoundary`. This is required in
+                cases where data is not oriented in the traditional way where
+                the first element of data (row 0, col 0) is the north-west
```
Comment (mraspaud) [inline pyresample/geometry.py:282]: An ascending pass and descending pass of a well behaved sensor (clockwise edges) have opposite first element orientations: north-west and south-east (not necessarily respectively, depending on how you look at it).
So I don't think north-west should be mentioned here, since the flip of lons and lats doesn't actually change the position of the first corner of the swath, right?

**Review by mraspaud — CHANGES_REQUESTED** [review]
Thanks for looking into this! I believe this could be made a bit more generic.

File: pyresample/geometry.py (hunk 1 above)
Comment (djhoese) [reply pyresample/geometry.py:984]: Left over from testing. I'll remove it.

File: pyresample/geometry.py (hunk 3 above)
Comment (djhoese) [reply pyresample/geometry.py:282]: What if I mentioned north-west and south-east?

File: pyresample/geometry.py (hunk 2 above)
Comment (djhoese) [reply pyresample/geometry.py:315]: Oh good point on the poles. So I could do the math on the first corner to get the angle between the two vectors in 3D space, but won't that always give the smallest angle between them? It's been a while since I've done this type of math (and I have a cold) but looking at https://www.omnicalculator.com/math/angle-between-two-vectors, how do I know if it is clockwise or counter-clockwise in 3D space?

File: pyresample/geometry.py (hunk 3 above)
Comment (mraspaud) [reply pyresample/geometry.py:282]: I think north-east and south-west cases can also happen (I'm thinking with Sentinel1 where I've seen this happen)

File: pyresample/geometry.py (hunk 2 above)
Comment (mraspaud) [reply pyresample/geometry.py:315]: because you are on the surface of a spheroid, you can give it an orientation...

Comment (mraspaud) [reply pyresample/geometry.py:315]: i.e. the normal of the angle has to point away from the centre of the earth

Comment (mraspaud) [reply pyresample/geometry.py:315]: maybe this can help? https://math.stackexchange.com/a/1715844

File: pyresample/geometry.py (hunk 3 above)
Comment (djhoese) [reply pyresample/geometry.py:282]: So you're saying that Sentinel1 data has either north-east or south-west as its first corner and the path followed by the existing logic still produces a clockwise series of coordinates? I suppose anything's possible. I'll see if I can remove the mention of north-west and be clearer about what I'm saying.

Comment (mraspaud) [reply pyresample/geometry.py:282]: yes exactly. The sentinel 1 data we use comes flipped horizontally or vertically depending whether the pass is ascending or descending...

File: pyresample/geometry.py [hunk 4]
```diff
@@ -269,9 +269,16 @@ def get_boundary_lonlats(self):
...
-    def get_bbox_lonlats(self) -> tuple:
+    def get_bbox_lonlats(self, force_clockwise: bool = True) -> tuple:
         """Return the bounding box lons and lats.
 
+        Args:
+            force_clockwise:
+                Perform minimal checks and reordering of coordinates to ensure
+                that the returned coordinates follow a clockwise direction.
+                This is important for compatibility with
+                :class:`pyresample.spherical.AreaBoundary`. Default is True.
```
Comment (mraspaud) [inline pyresample/geometry.py:280]: I think we need more explanation on why this is important, eg `A clockwise orientation of the boundaries ensures that a the spherical bounding box covers less than half of the globe, hence that a regular swath bounding box is oriented correctly with regards to defining its inside and outside.`

File: pyresample/geometry.py [hunk 5]
```diff
@@ -298,8 +305,29 @@ def get_bbox_lonlats(self) -> tuple:
...
+    def _corner_is_clockwise(lon1, lat1, corner_lon, corner_lat, lon2, lat2):
+        from pyresample.spherical import Arc, SCoordinate
+        point1 = SCoordinate(math.radians(lon1), math.radians(lat1))
+        point2 = SCoordinate(math.radians(corner_lon), math.radians(corner_lat))
+        point3 = SCoordinate(math.radians(lon2), math.radians(lat2))
+        arc1 = Arc(point1, point2)
+        arc2 = Arc(point2, point3)
+        # a straight line has an angle of 0
+        # if the second segment is clockwise the angle will go negative
+        # if the second segment is counter-clockwise the angle will go positive
```
Comment (mraspaud) [inline pyresample/geometry.py:327]: Should we move this to the docstring?

File: pyresample/geometry.py [hunk 6]
```diff
@@ -298,8 +305,29 @@ def get_bbox_lonlats(self) -> tuple:
...
+        angle = arc1.angle(arc2)
+        return -np.pi < angle < 0
```
Comment (mraspaud) [inline pyresample/geometry.py:329]: Just for extra clarity
```suggestion
        is_clockwise = -np.pi < angle < 0
        return is_clockwise
```

File: pyresample/geometry.py [hunk 7]
```diff
@@ -1050,8 +1078,10 @@ def _compute_bound_centers_dask(self, proj_dict, lons, lats):
         y_is_pole = (ymax >= 90 - epsilon) or (ymin <= -90 + epsilon)
         if crs.is_geographic and x_passes_antimeridian and not y_is_pole:
```
Comment (mraspaud) [inline pyresample/geometry.py:1079]: If the crs is geographic, I would rather go to cartesian coordinates to do the computations here... but maybe this should be another PR?

File: pyresample/test/utils.py [hunk 8]
```diff
@@ -166,7 +166,8 @@ def __exit__(self, type, value, traceback):
 
 def create_test_longitude(start, stop, shape, twist_factor=0.0, dtype=np.float32):
     """Get basic sample of longitude data."""
-    if start > stop:
+    if start > stop and start > 0 > stop:
```
Comment (mraspaud) [inline pyresample/test/utils.py:169]: doesn't the second one imply the first one? ie would it be enough with
`if start > 0 > stop:` ?

File: pyresample/geometry.py [hunk 9]
```diff
@@ -298,8 +305,29 @@ def get_bbox_lonlats(self) -> tuple:
                            (s4_lon.squeeze(), s4_lat.squeeze())])
         if hasattr(lons[0], 'compute') and da is not None:
             lons, lats = da.compute(lons, lats)
+        if force_clockwise and not self._corner_is_clockwise(
+                lons[0][-2], lats[0][-2], lons[0][-1], lats[0][-1], lons[1][1], lats[1][1]):
+            # going counter-clockwise
+            # swap the side order and the order of the values in each side
+            # to make it clockwise
+            lons = [lon[::-1] for lon in lons[::-1]]
+            lats = [lat[::-1] for lat in lats[::-1]]
```
Comment (mraspaud) [inline pyresample/geometry.py:314]: Should this be it's own function to clarify the code and not need the comments above?

**Review by mraspaud — CHANGES_REQUESTED** [review]
Good work, just some minor notes

File: pyresample/geometry.py (hunk 4 above)
Comment (djhoese) [reply pyresample/geometry.py:280]: Is this the reason why? I wouldn't know as the spherical geometry stuff isn't documented so I just assumed it has to be done "because of the math". If the coordinates were counter-clockwise but the math supported/expected that then it should be fine, right? The main "issue" is that the spherical geometry expects clockwise so it must be clockwise.

File: pyresample/geometry.py (hunk 5 above)
Comment (djhoese) [reply pyresample/geometry.py:327]: Yes, probably.

File: pyresample/geometry.py (hunk 7 above)
Comment (djhoese) [reply pyresample/geometry.py:1079]: Yes probably another PR. However, I'm not sure why cartesian coordinates would work better for this math. It is meant to be a simple workaround so my hope was that it wouldn't get too complicated. But if you see something then of course we could do it.

File: pyresample/geometry.py (hunk 9 above)
Comment (djhoese) [reply pyresample/geometry.py:314]: Any good name suggestions for the function?

File: pyresample/test/utils.py (hunk 8 above)
Comment (djhoese) [reply pyresample/test/utils.py:169]: Probably. This was in the middle of testing and I just wanted to make sure I had the right guess as to what was going wrong. I'll clean it up in a little while.

File: pyresample/geometry.py (hunk 4 above)
Comment (mraspaud) [reply pyresample/geometry.py:280]: it has to go clockwise around the inside, because that's how we define inside and outside. What we want is to have the inside were the data is...

File: pyresample/geometry.py (hunk 7 above)
Comment (mraspaud) [reply pyresample/geometry.py:1079]: it's just that I don't like the thresholding needed around poles and antimeridian...

File: pyresample/geometry.py (hunk 9 above)
Comment (mraspaud) [reply pyresample/geometry.py:314]: `reverse_boundaries` ?

File: pyresample/geometry.py (hunk 4 above)
Comment (djhoese) [reply pyresample/geometry.py]: I made an attempt. Let me know what you think. I forgot that you made your initial suggestion for what it could say, but I still stand by the point that it doesn't matter "covers less than half the globe" as this is all specific to how the spherical geometry stuff is defined.

**Review by mraspaud — APPROVED** [review]
LGTM
